src/Common/src/Microsoft/Win32/SafeHandles/GssSafeHandles.cs

```diff
@@ -0,0 +1,158 @@
+                Interop.NetSecurityNative.CopyBuffer(this, destination, capacity: destination.Length, offset: offset);
+            }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ReleaseBuffer(out minorStatus, handle);
+            SetHandle(IntPtr.Zero);
+            return status == Interop.NetSecurityNative.Status.GSS_S_COMPLETE;
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a gss_name_t_desc*
+    /// </summary>
+    internal sealed class SafeGssNameHandle : SafeHandle
+    {
+        public static SafeGssNameHandle Create(string name, bool isUser)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name), "Invalid name passed to SafeGssNameHandle create");
+            SafeGssNameHandle retHandle;
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = isUser?
+                Interop.NetSecurityNative.ImportUserName(out minorStatus, name, name.Length, out retHandle) :
+                Interop.NetSecurityNative.ImportPrincipalName(out minorStatus, name, name.Length, out retHandle);
+            if (status != Interop.NetSecurityNative.Status.GSS_S_COMPLETE)
+            {
+                throw new Interop.NetSecurityNative.GssApiException(status, minorStatus);
+            }
+
+            return retHandle;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ReleaseName(out minorStatus, ref handle);
+            SetHandle(IntPtr.Zero);
+            return status == Interop.NetSecurityNative.Status.GSS_S_COMPLETE;
+        }
+
+        private SafeGssNameHandle()
+            : base(IntPtr.Zero, true)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a gss_cred_id_t_desc_struct*
+    /// </summary>
+    internal class SafeGssCredHandle : SafeHandle
+    {
+        public static SafeGssCredHandle Create(string username, string password, string domain)
+        {
+            SafeGssCredHandle retHandle = null;
+
+            // Empty username is OK if Kerberos ticket was already obtained
+            if (!string.IsNullOrEmpty(username))
+            {
+                using (SafeGssNameHandle userHandle = SafeGssNameHandle.Create(username, true))
+                {
+                    Interop.NetSecurityNative.Status status;
+                    Interop.NetSecurityNative.Status minorStatus;
+                    if (string.IsNullOrEmpty(password))
+                    {
+                        status = Interop.NetSecurityNative.AcquireCredSpNego(out minorStatus, userHandle, true, out retHandle);
```

Comment:
After searching for AcquireCredSpNego, I now know that the "true" here is `isInitiate`.  Please consider changing the interop signature from AcquireCredSpNego(..., bool, ...) to InitiateCredSpNego and AcceptCredSpNego, to make it more clear to a reader what's going on.